{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "colab_type": "text",
    "id": "view-in-github"
   },
   "source": [
    "<a href=\"https://colab.research.google.com/github/yc386/orthrus_metaproteomics/blob/main/annotated_orthrus_pt2.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "g529CpYsDdjc"
   },
   "source": [
    "# **Orthrus pt2- [`Sage`](https://github.com/lazear/sage) + [`Mokapot`](https://github.com/wfondrie/mokapot)**\n",
    "\n",
    "Please note: change to **TPU** runtime if RAM usage is expected to be high (due to files/PTMs/databases)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "QSA5zC6q6knU"
   },
   "outputs": [],
   "source": [
    "#@title Add inputs for `SAGE` -> click `Runtime` -> `Run all`\n",
    "\n",
    "#@markdown **Parameters for `SAGE`**\n",
    "peak_folder = \"data/PXD027613/mzML\"  #@param {type:\"string\"}\n",
    "file_type = \"mzML\"  #@param [\"mzML\", \"mgf\"]\n",
    "#@markdown - use the drop-down menu to choose the instrument file type\n",
    "\n",
    "#@markdown **_De novo_ peptide sequencing algorithm**\n",
    "algorithm = \"instanovo\"  #@param [\"instanovo\", \"casanovo\"]\n",
    "\n",
    "#@markdown **Option 1: `SAGE` PTM plus**\n",
    "#@markdown - Default `Sage` contains\n",
    "use_PTM_plus = True  #@param {type:\"boolean\"}\n",
    "max_variable_mods = 3  #@param {type:\"number\"}\n",
    "missed_cleavages = 2  #@param {type:\"number\"}\n",
    "AA_1 = \"M\"  #@param [\"None\", \"[\",\"]\",\"A\", \"C\", \"D\", \"E\", \"F\", \"G\", \"H\", \"I\", \"K\", \"L\", \"M\", \"N\", \"P\", \"Q\", \"R\", \"S\", \"T\", \"V\", \"W\", \"Y\"]\n",
    "AA_1_mod = 15.9949  #@param {type:\"number\"}\n",
    "AA_2 = \"P\"  #@param [\"None\", \"[\",\"]\",\"A\", \"C\", \"D\", \"E\", \"F\", \"G\", \"H\", \"I\", \"K\", \"L\", \"M\", \"N\", \"P\", \"Q\", \"R\", \"S\", \"T\", \"V\", \"W\", \"Y\"]\n",
    "AA_2_mod = 15.9949  #@param {type:\"number\"}\n",
    "AA_3 = \"N\"  #@param [\"None\", \"[\",\"]\",\"A\", \"C\", \"D\", \"E\", \"F\", \"G\", \"H\", \"I\", \"K\", \"L\", \"M\", \"N\", \"P\", \"Q\", \"R\", \"S\", \"T\", \"V\", \"W\", \"Y\"]\n",
    "AA_3_mod = 0.984016  #@param {type:\"number\"}\n",
    "AA_4 = \"Q\"  #@param [\"None\", \"[\",\"]\",\"A\", \"C\", \"D\", \"E\", \"F\", \"G\", \"H\", \"I\", \"K\", \"L\", \"M\", \"N\", \"P\", \"Q\", \"R\", \"S\", \"T\", \"V\", \"W\", \"Y\"]\n",
    "AA_4_mod = 0.984016  #@param {type:\"number\"}\n",
    "AA_5 = \"None\"  #@param [\"None\", \"[\",\"]\",\"A\", \"C\", \"D\", \"E\", \"F\", \"G\", \"H\", \"I\", \"K\", \"L\", \"M\", \"N\", \"P\", \"Q\", \"R\", \"S\", \"T\", \"V\", \"W\", \"Y\"]\n",
    "#@markdown - [ = n-terminal\n",
    "AA_5_mod = 42.010565  #@param {type:\"number\"}\n",
    "\n",
    "#@markdown **Option 2: `Mokapot` rescoring**\n",
    "use_Mokapot = True  #@param {type:\"boolean\"}\n",
    "#@markdown - machine learning-based rescoring per experiment or across experiments\n",
    "joint_modelling = False  #@param {type:\"boolean\"}\n",
    "#@markdown - a joint model for low abundance samples\n",
    "default_Percolator = False  #@param {type:\"boolean\"}\n",
    "#@markdown - Python implementation of the Percolator SVM model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "6KE6TaQj3oxE",
    "outputId": "a9f4c213-f21c-4d8f-f9b0-b1af41a14092"
   },
   "outputs": [],
   "source": [
    "#@title install dependencies\n",
    "\n",
    "import os\n",
    "import shutil\n",
    "import subprocess\n",
    "import glob\n",
    "import json\n",
    "\n",
    "\n",
    "def run_command(command: str):\n",
    "    \"\"\"Runs a shell command and streams its output in real-time.\"\"\"\n",
    "    process = subprocess.Popen(\n",
    "        command,\n",
    "        stdout=subprocess.PIPE,\n",
    "        stderr=subprocess.PIPE,\n",
    "        shell=True,  # Enables the use of a string command\n",
    "        text=True,  # Ensures output is in string format\n",
    "    )\n",
    "\n",
    "    # Stream the output\n",
    "    while True:\n",
    "        output = process.stdout.readline()  # type: ignore\n",
    "        if output == \"\" and process.poll() is not None:\n",
    "            break\n",
    "        if output:\n",
    "            print(output.strip())\n",
    "\n",
    "    # Check for errors\n",
    "    stderr = process.stderr.read()  # type: ignore\n",
    "    if stderr:\n",
    "        print(f\"Error: {stderr.strip()}\")\n",
    "\n",
    "    # Return the exit code\n",
    "    return process.returncode\n",
    "\n",
    "\n",
    "def organise_files(directory, algorithm):\n",
    "    \"\"\"Organise mztabs and instrument files in the same folder.\"\"\"\n",
    "    if not os.path.isdir(directory):\n",
    "        print(f\"The directory {directory} does not exist.\")\n",
    "        return\n",
    "\n",
    "    mzml_files = glob.glob(os.path.join(directory, \"*.mzML\"))\n",
    "\n",
    "    for mzml_file in mzml_files:\n",
    "        base_name = os.path.splitext(os.path.basename(mzml_file))[0]\n",
    "        core_name = base_name.split(\"_\")[\n",
    "            0\n",
    "        ]  # Extract the part before the de novo algorithm\n",
    "\n",
    "        new_folder_path = os.path.join(directory, base_name)\n",
    "\n",
    "        if not os.path.exists(new_folder_path):\n",
    "            os.makedirs(new_folder_path)\n",
    "            print(f\"Created folder: {new_folder_path}\")\n",
    "        else:\n",
    "            print(f\"Folder already exists: {new_folder_path}\")\n",
    "\n",
    "        new_mzml_path = os.path.join(new_folder_path, os.path.basename(mzml_file))\n",
    "        if not os.path.exists(new_mzml_path):\n",
    "            shutil.move(mzml_file, new_folder_path)\n",
    "            print(f\"Moved {mzml_file} to {new_folder_path}\")\n",
    "        else:\n",
    "            print(f\".mzML file already exists in the destination: {new_mzml_path}\")\n",
    "\n",
    "        fasta_filename = f\"{core_name}_{algorithm}_matched.fasta\"\n",
    "        fasta_file = os.path.join(directory, fasta_filename)\n",
    "\n",
    "        if os.path.exists(fasta_file):\n",
    "            new_fasta_path = os.path.join(new_folder_path, fasta_filename)\n",
    "            if not os.path.exists(new_fasta_path):\n",
    "                shutil.move(fasta_file, new_folder_path)\n",
    "                print(f\"Moved {fasta_file} to {new_folder_path}\")\n",
    "            else:\n",
    "                print(\n",
    "                    f\".fasta file already exists in the destination: {new_fasta_path}\"\n",
    "                )\n",
    "        else:\n",
    "            print(f\"No matching .fasta file found for {base_name}\")\n",
    "\n",
    "\n",
    "def get_sage_config(\n",
    "    json_file_path,\n",
    "    peak_folder,\n",
    "    static_mods,\n",
    "    new_mods,\n",
    "    missed_cleavages,\n",
    "    min_len,\n",
    "    max_len,\n",
    "    max_variable_mods,\n",
    "    output_config_path,\n",
    "):\n",
    "    \"\"\"Update the Sage config file with the new parameters.\"\"\"\n",
    "    with open(json_file_path, \"r\") as file:\n",
    "        json_data = json.load(file)\n",
    "\n",
    "        peak_files = glob.glob(peak_folder)\n",
    "        print(f\"üóÇÔ∏è {len(peak_files)} file(s) collected from {peak_folder}\")\n",
    "        json_data[\"mzml_paths\"] = peak_files\n",
    "        json_data[\"database\"][\"static_mods\"] = static_mods\n",
    "        json_data[\"database\"][\"variable_mods\"] = new_mods\n",
    "        json_data[\"database\"][\"enzyme\"][\"missed_cleavages\"] = missed_cleavages\n",
    "        json_data[\"database\"][\"enzyme\"][\"min_len\"] = min_len\n",
    "        json_data[\"database\"][\"enzyme\"][\"max_len\"] = max_len\n",
    "        json_data[\"database\"][\"max_variable_mods\"] = max_variable_mods\n",
    "        json_data[\"database\"][\"decoy_tag\"] = \"rev_\"\n",
    "        json_data[\"database\"][\"generate_decoys\"] = True\n",
    "\n",
    "    with open(output_config_path, \"w\") as f:\n",
    "        json.dump(json_data, f, indent=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "qDo9CbOsBWIU",
    "outputId": "83c3eb3f-2453-4470-911b-52c2a12964e3"
   },
   "outputs": [],
   "source": [
    "#@title Run Sage\n",
    "\n",
    "organise_files(peak_folder, algorithm)\n",
    "folder_path = peak_folder\n",
    "\n",
    "if use_PTM_plus:\n",
    "    AAs = [AA_1, AA_2, AA_3, AA_4, AA_5]\n",
    "    mods = [AA_1_mod, AA_2_mod, AA_3_mod, AA_4_mod, AA_5_mod]\n",
    "    PTMs = {}\n",
    "    for AA, mod in zip(AAs, mods):\n",
    "        if AA != \"None\":\n",
    "            PTMs[AA] = [mod]\n",
    "    big_folder = glob.glob(f\"{folder_path}/*\")\n",
    "    for folder in big_folder:\n",
    "        if not os.path.isdir(folder):\n",
    "            continue\n",
    "        mzml_files = glob.glob(f\"{folder}/*.{file_type}\")\n",
    "        peak_path = mzml_files[0]\n",
    "        output_json = peak_path.replace(f\".{file_type}\", \".json\")\n",
    "        json_file_path = \"config.json\"\n",
    "        missed_cleavages = missed_cleavages\n",
    "        min_len = 6\n",
    "        max_len = 30\n",
    "        max_variable_mods = max_variable_mods\n",
    "        static_mods = {\"C\": 57.021464}\n",
    "        get_sage_config(\n",
    "            json_file_path,\n",
    "            peak_path,\n",
    "            static_mods,\n",
    "            PTMs,\n",
    "            missed_cleavages,\n",
    "            min_len,\n",
    "            max_len,\n",
    "            max_variable_mods,\n",
    "            output_json,\n",
    "        )\n",
    "        fasta_files = glob.glob(f\"{folder}/*.fasta\")\n",
    "        fasta_path = fasta_files[0]\n",
    "        run_command(\n",
    "            f\"sage {output_json} --fasta {fasta_path} --write-pin --output_directory {folder}\"\n",
    "        )\n",
    "\n",
    "else:\n",
    "    big_folder = glob.glob(f\"{folder_path}/*\")\n",
    "    for folder in big_folder:\n",
    "        if not os.path.isdir(folder):\n",
    "            continue\n",
    "        mzml_files = glob.glob(f\"{folder}/*.{file_type}\")\n",
    "        peak_path = mzml_files[0]\n",
    "        output_json = peak_path.replace(f\".{file_type}\", \".json\")\n",
    "        json_file_path = \"config.json\"\n",
    "        missed_cleavages = 2\n",
    "        min_len = 6\n",
    "        max_len = 30\n",
    "        max_variable_mods = 5\n",
    "        static_mods = {\"C\": 57.021464}\n",
    "        new_mods = {\"M\": [15.994915], \"N\": [0.984016], \"Q\": [0.984016]}\n",
    "        get_sage_config(\n",
    "            json_file_path,\n",
    "            peak_path,\n",
    "            static_mods,\n",
    "            new_mods,\n",
    "            missed_cleavages,\n",
    "            min_len,\n",
    "            max_len,\n",
    "            max_variable_mods,\n",
    "            output_json,\n",
    "        )\n",
    "        fasta_files = glob.glob(f\"{folder}/*.fasta\")\n",
    "        fasta_path = fasta_files[0]\n",
    "        run_command(\n",
    "            f\"sage {output_json} --fasta {fasta_path} --write-pin --output_directory {folder}\"\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "vGSAYJN1_923",
    "outputId": "1a3662e3-ca81-4459-f7c5-93f7c795225a"
   },
   "outputs": [],
   "source": [
    "#@title Brew Mokapot\n",
    "\n",
    "import mokapot\n",
    "from xgboost import XGBClassifier\n",
    "from sklearn.model_selection import GridSearchCV\n",
    "import numpy as np\n",
    "\n",
    "# XGBoost schema from Fondrie & Noble 2021). A non-linear XGBoost seems to be better for rescoring open search results.\n",
    "\n",
    "grid = {\n",
    "    \"scale_pos_weight\": np.logspace(0, 2, 3),\n",
    "    \"max_depth\": [1, 3, 6],\n",
    "    \"min_child_weight\": [1, 10, 100],\n",
    "    \"gamma\": [0, 1, 10],\n",
    "}\n",
    "xgb_mod = GridSearchCV(\n",
    "    XGBClassifier(),\n",
    "    param_grid=grid,\n",
    "    n_jobs=1,\n",
    "    cv=3,\n",
    "    scoring=\"roc_auc\",\n",
    ")\n",
    "\n",
    "\n",
    "def get_all_pin_files(folder_path):\n",
    "    \"\"\"Get all .pin files in the folder.\"\"\"\n",
    "    psm_files = []\n",
    "    for root, _dirs, files in os.walk(folder_path):\n",
    "        for file in files:\n",
    "            if file.endswith(\".pin\"):\n",
    "                full_path = os.path.join(root, file)\n",
    "                psm_files.append(full_path)\n",
    "    return psm_files\n",
    "\n",
    "\n",
    "if use_Mokapot:\n",
    "    folder_path = peak_folder\n",
    "\n",
    "    if joint_modelling:\n",
    "        psm_files = get_all_pin_files(peak_folder)\n",
    "        if default_Percolator:\n",
    "            svm = mokapot.PercolatorModel()\n",
    "            psm_list = mokapot.read_pin(psm_files)\n",
    "            results, models = mokapot.brew(psm_list, svm)\n",
    "            result_files = results.to_txt(peak_folder)\n",
    "        else:\n",
    "            mod = mokapot.Model(xgb_mod)\n",
    "            psm_list = mokapot.read_pin(psm_files)\n",
    "            results, models = mokapot.brew(psm_list, mod)\n",
    "            result_files = results.to_txt(peak_folder)\n",
    "\n",
    "    else:\n",
    "        big_folder = sorted(glob.glob(f\"{folder_path}/*\"))\n",
    "        for folder in big_folder:\n",
    "            if not os.path.isdir(folder):\n",
    "                continue\n",
    "            print(f\"Processing folder: {folder}\")\n",
    "            pin_files = glob.glob(f\"{folder}/*.pin\")\n",
    "            if not pin_files:\n",
    "                print(f\"No .pin files found in {folder}. Skipping...\")\n",
    "                continue\n",
    "            pin = pin_files[0]\n",
    "            if default_Percolator:\n",
    "                svm = mokapot.PercolatorModel()\n",
    "                psm_list = mokapot.read_pin(pin)\n",
    "                results, models = mokapot.brew(psm_list, svm)\n",
    "                result_files = results.to_txt(folder)\n",
    "            else:\n",
    "                mod = mokapot.Model(xgb_mod)\n",
    "                psm_list = mokapot.read_pin(pin)\n",
    "                results, models = mokapot.brew(psm_list, mod)\n",
    "                result_files = results.to_txt(folder)\n",
    "\n",
    "else:\n",
    "    print(\"Mokapot not brewed\")"
   ]
  }
 ],
 "metadata": {
  "accelerator": "TPU",
  "colab": {
   "authorship_tag": "ABX9TyOc6na9FKQ3IQAJawdwwn0R",
   "gpuType": "V28",
   "include_colab_link": true,
   "machine_shape": "hm",
   "mount_file_id": "1wFx74CFxO_qDFRD8MfDxqWJ3ZiaOkBXA",
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3",
   "name": "python3"
  },
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
